Migrate docker publication GitLab jobs to public-images

L3n41c opened the pull request:
### What does this PR do?

Migrate the GitLab jobs to publish docker images to [DataDog/public-images](https://github.com/DataDog/public-images).

### Motivation

`public-images` will centralise all the complexity of building, pushing and signing multi-arch images on several repositories.

### Additional Notes

The child job will be in charge of deploying the docker image on Docker Hub, GCR and ECR [unless IMG_REGISTRIES is set to `dev`, in which case the image will only be published on Docker Hub](https://github.com/DataDog/public-images/pull/9/files).
For dev images, we want to deploy them only on Docker Hub to reduce the cost.
For all release images, we want to deploy them on all our registries.

### Describe how to test your changes

Try to publish docker images and check the content of docker hub and gcr.io.

L3n41c committed:
Migrate docker publication GitLab jobs to public-images

KSerrania commented on .gitlab-ci.yml:
```diff
@@ -290,9 +293,17 @@ variables:
 .on_deploy_a6_manual:
   - <<: *if_not_version_6
     when: never
+  - <<: *if_tagged_deploy
+    when: manual
+    allow_failure: true
+    variables:
+      AGENT_REPOSITORY: agent
   - <<: *if_deploy
     when: manual
     allow_failure: true
+    variables:
+      AGENT_REPOSITORY: agent-dev
+      IMG_REGISTRIES: dev
```
The previous behavior was the following: if the release branch is `nightly`, send to `agent-dev`, otherwise send to `agent`: https://github.com/DataDog/datadog-agent/blob/bea56884ae152e63f1aea5d726a70f69a3234544/.gitlab/deploy_7/docker.yml#L57-L59

I think we want to keep that. To do this, you can do:
```suggestion
.on_deploy_a6_manual:
  - <<: *if_not_version_6
    when: never
  - <<: *if_not_deploy
    when: never
  - <<: *if_not_stable_or_beta_repo_branch
    when: manual
    allow_failure: true
    variables:
      AGENT_REPOSITORY: agent-dev
      IMG_REGISTRIES: dev
  - when: manual
    allow_failure: true
    variables:
      AGENT_REPOSITORY: agent
```

You'll need to define:
```
.if_not_deploy: &if_not_deploy
  if: $DEPLOY_AGENT != "true"
```

instead of `.if_tagged_commit`

KSerrania commented on .gitlab-ci.yml:
```diff
@@ -302,9 +313,19 @@ variables:
 .on_deploy_a7_manual:
   - <<: *if_not_version_7
     when: never
+  - <<: *if_tagged_deploy
+    when: manual
+    allow_failure: true
+    variables:
+      AGENT_REPOSITORY: agent
+      DSD_REPOSITORY: dogstatsd
   - <<: *if_deploy
     when: manual
     allow_failure: true
+    variables:
+      AGENT_REPOSITORY: agent-dev
+      DSD_REPOSITORY: dogstatsd-dev
+      IMG_REGISTRIES: dev
```
Same here:

```suggestion
.on_deploy_a7_manual:
  - <<: *if_not_version_7
    when: never
  - <<: *if_not_deploy
    when: never
  - <<: *if_not_stable_or_beta_repo_branch
    when: manual
    allow_failure: true
    variables:
      AGENT_REPOSITORY: agent-dev
      DSD_REPOSITORY: dogstatsd-dev
      IMG_REGISTRIES: dev
  - when: manual
    allow_failure: true
    variables:
      AGENT_REPOSITORY: agent
      DSD_REPOSITORY: dogstatsd
```

L3n41c committed:
gitlab-ci: Define `if_not_deploy` instead of `if_tagged_deploy`

L3n41c committed:
Update .gitlab-ci.yml

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

L3n41c committed:
Update .gitlab-ci.yml

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

L3n41c commented:
[Here is an example of pipeline created with `RUN_ALL_BUILDS = true`](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/4749867).
I manually launched the docker image deployment jobs.
We can check the published multi-arch images there:
* https://hub.docker.com/repository/docker/datadog/agent-dev/tags?page=1&ordering=last_updated&name=lenaic-public-images

L3n41c committed:
Do not sign docker images in the `*-scan` repositories

Those private repositories are used for vulnerability scanners only.
Nobody is expected to pull from them.
That’s why notary isn’t setup on them.

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+    strategy: depend
+  parallel:
+    matrix:
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-amd64
```
We don't have `2009` images, only `2004` ones:
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2004-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+    matrix:
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:${VERSION}
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2004-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:${VERSION}
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:${VERSION}-jmx
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-servercore-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2004-servercore-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:${VERSION}-jmx
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-servercore-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:${VERSION}-servercore
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-servercore-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2004-servercore-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+    strategy: depend
+  parallel:
+    matrix:
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2004-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+    matrix:
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:7,${AGENT_REPOSITORY}:latest
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2004-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:7,${AGENT_REPOSITORY}:latest
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-arm64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:7-jmx,${AGENT_REPOSITORY}:latest-jmx
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-servercore-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2004-servercore-amd64
```

KSerrania commented on .gitlab/deploy_7/docker.yml:
```diff
@@ -6,319 +6,77 @@
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:7-jmx,${AGENT_REPOSITORY}:latest-jmx
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-win2009-servercore-amd64
+        IMG_DESTINATIONS: ${AGENT_REPOSITORY}:7-servercore,${AGENT_REPOSITORY}:latest-servercore
+      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2009-servercore-amd64
```
```suggestion
      - IMG_SOURCES: ${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1809-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win1909-servercore-amd64,${SRC_AGENT}:v$CI_PIPELINE_ID-${CI_COMMIT_SHORT_SHA}-7-jmx-win2004-servercore-amd64
```

L3n41c committed:
Apply suggestions from code review

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

L3n41c committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into lenaic/public-images

L3n41c committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into lenaic/public-images

L3n41c committed:
Surround `CI_PIPELINE_ID` by curly brackets

L3n41c committed:
Use `inv pipeline.trigger-child-pipeline` to be able to use `inv agent.version`

albertvaka committed:
Merge remote-tracking branch 'origin/main' into lenaic/public-images

albertvaka commented:
Test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/4870298

L3n41c commented:
@albertvaka Here are the images published by the test pipeline: https://hub.docker.com/repository/docker/datadog/agent-dev/tags?page=1&ordering=last_updated&name=nightly-8d751710

albertvaka committed:
Add missing image field to jobs that have a script

albertvaka commented:
New test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/pipelines/4877760

albertvaka committed:
Add missing tags field

albertvaka commented:
New test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/pipelines/4888964

albertvaka committed:
Trigger jobs don't need dependencies

albertvaka committed:
Turn script into array

Otherwise it concatenates everything into a single command

albertvaka commented:
New test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/pipelines/4890071

albertvaka committed:
Missing "export"

albertvaka commented:
New test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/pipelines/4891320

albertvaka committed:
Fix typo in image source name

albertvaka commented:
New test pipeline: https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/4892496

L3n41c committed:
Fix bug in publication of `latest-py2` images

L3n41c committed:
Add forgotten dependencies directive for deploy jobs

albertvaka committed:
Replace `trigger` gitlab keyword by `inv pipeline.trigger-child-pipeline`

albertvaka committed:
Use trigger-child-pipeline in all manual jobs

albertvaka committed:
Merge remote-tracking branch 'origin/main' into lenaic/public-images

albertvaka commented:
Test pipeline (skipping kitchen tests): https://gitlab.ddbuild.io/DataDog/datadog-agent/pipelines/4906046

L3n41c committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into lenaic/public-images
